dids/didjwk.go

```diff
@@ -0,0 +1,117 @@
+package dids
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+
+	"github.com/tbd54566975/web5-go/crypto"
+	"github.com/tbd54566975/web5-go/crypto/dsa"
+	"github.com/tbd54566975/web5-go/jwk"
+)
+
+type newDIDJWKOptions struct {
+	keyManager  crypto.KeyManager
+	algorithmID string
+}
+
+type newDIDJWKOption func(o *newDIDJWKOptions)
+
+// KeyManager is an option that can be passed to NewDIDJWK to provide a KeyManager
+func KeyManager(k crypto.KeyManager) newDIDJWKOption {
+	return func(o *newDIDJWKOptions) {
+		o.keyManager = k
+	}
+}
+
+// AlgorithmID is an option that can be passed to NewDIDJWK to specify a specific
+// cryptographic algorithm to use to generate the private key
+func AlgorithmID(id string) newDIDJWKOption {
+	return func(o *newDIDJWKOptions) {
+		o.algorithmID = id
+	}
+}
+
+// NewDIDJWK can be used to generate a new `did:jwk`. `did:jwk` is useful in scenarios where:
+//   - Offline resolution is preferred
+//   - Key rotation is not required
+//   - Service endpoints are not necessary
+//
+// Spec: https://github.com/quartzjer/did-jwk/blob/main/spec.md
+func NewDIDJWK(opts ...newDIDJWKOption) (BearerDID, error) {
+	o := newDIDJWKOptions{
+		keyManager:  crypto.NewLocalKeyManager(),
+		algorithmID: dsa.AlgorithmID.ED25519,
+	}
+
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	keyMgr := o.keyManager
+
+	keyID, err := keyMgr.GeneratePrivateKey(o.algorithmID)
+	if err != nil {
+		return BearerDID{}, fmt.Errorf("failed to generate private key: %w", err)
+	}
+
+	publicJWK, _ := keyMgr.GetPublicKey(keyID)
+	bytes, err := json.Marshal(publicJWK)
+	if err != nil {
+		return BearerDID{}, fmt.Errorf("failed to marshal public key: %w", err)
+	}
+
+	id := base64.RawURLEncoding.EncodeToString(bytes)
+	did := BearerDID{
+		DID: DID{
+			Method: "jwk",
+			URI:    "did:jwk:" + id,
+			ID:     id,
+		},
+		KeyManager: keyMgr,
+	}
+
+	return did, nil
+}
+
+// Resolves the provided DID URI
+func ResolveDIDJWK(uri string) ResolutionResult {
+	did, err := ParseURI(uri)
+	if err != nil {
+		return ResolutionResultWithError("invalidDid")
+	}
+
+	if did.Method != "jwk" {
+		return ResolutionResultWithError("invalidDid")
+	}
+
+	decodedID, err := base64.RawURLEncoding.DecodeString(did.ID)
+	if err != nil {
+		return ResolutionResultWithError("invalidDid")
+	}
+
+	var jwk jwk.JWK
+	err = json.Unmarshal(decodedID, &jwk)
```

Comment:
Unfortunately the default `json.Umarshal()` allows additional parameters. For something like this where you want as much correctness as possible I would suggest instantiating a `json.Decoder` and calling [DisallowUnknownFields](https://cs.opensource.google/go/go/+/go1.21.6:src/encoding/json/stream.go;l=42) on it.

Reply:
I should caveat this by saying that `DisallowUnknownFields` isn't as future compatible. That is, if you add additional fields to JWK in future, it would fail.